Pal 869 user service auditing

developer01189998819991197253 opened the pull request.

developer01189998819991197253 commented:
After PR has been approved I will squash and recommit to remove unsigned commits

kg981167224 commented on user-service/README.md:
```diff
@@ -1,26 +1,104 @@
-<span style="color:red">**Note:** As noted in the [documentation root](../README.md) Palisade is in an early stage of development, therefore the precise APIs contained in these service modules are likely to adapt as the project matures.</span>
+The User Service is responsible for providing the other Palisade components with knowledge of the users of the system. 
+Note that we are avoiding having a separate notion of "Palisade" users that are distinct from the normal user accounts of the system. There is no such thing as a "Palisade" user.  
+One of the purposes of this service is to allow Palisade to adopt the whatever notion of *user* the host environment has. For example, this may from a central directory service such as LDAP, 
```
```suggestion
One of the purposes of this service is to allow Palisade to adopt whatever notion of *user* the host environment has. For example, this may from a central directory service such as LDAP, 
```

kg981167224 commented on user-service/src/component-tests/java/uk/gov/gchq/palisade/component/user/service/CaffeineUserCachingTest.java:
```diff
@@ -65,91 +62,79 @@ private void forceCleanUp() {
     }
 
     @Test
     void testNonExistentUserRetrieveFails() {
```
Is this test reliable? Since there is a sleep for 1s and TTL is 1s isn't there a chance the cache doesn't get cleaned up, especially if the sleep gets an interrupt.

developer01189998819991197253 replied:
the testNonExistentUserRetrieveFails test never adds a user to the cache, so gets a user that never exists and then throws the exception

kg981167224 commented on user-service/src/component-tests/java/uk/gov/gchq/palisade/component/user/service/CaffeineUserCachingTest.java:
```diff
@@ -65,91 +62,79 @@ private void forceCleanUp() {
         userService.addUser(user);
 
-        TimeUnit.MILLISECONDS.sleep(1000);
+        TimeUnit.SECONDS.sleep(1);
```
See earlier question on sleep.

developer01189998819991197253 replied:
The test has been reliable so far, but also below runs the forceCleanUp() method that uses the cacheManager to clean up the cache manually if the sleep gets interrupted etc

kg981167224 commented on user-service/src/component-tests/java/uk/gov/gchq/palisade/component/user/service/RedisUserCachingTest.java:
```diff
@@ -0,0 +1,180 @@
+
+        private static final int REDIS_PORT = 6379;
+
+        static final GenericContainer<?> REDIS = new GenericContainer<>("redis:6-alpine")
```
Is this going to pull this image every time? Or will it use a local cache. Maybe worth testing to ensure we aren't going to hit any rate issues. Especially when run in pipeline

developer01189998819991197253 replied:
it should use locally cached images

dev930018 replied:
Also, any rate issues would only be against redis's pull allowance, not ours, no?

kg981167224 replied:
I think any test container pull will do the request to DH logging our IP address with docker hub, which will come off our allocation. TBH If it uses local cache won't be a problem.

dbz120896 replied:
When run in Jenkins, this will indeed pull the container every time. We need to do the same as we have planned for Maven, and have a EFS volume which is shared.

kg981167224 commented on user-service/src/component-tests/java/uk/gov/gchq/palisade/component/user/service/RedisUserCachingTest.java:
```diff
@@ -0,0 +1,180 @@
+
+        // When - we try to access stale cache data
+        Exception noSuchUserId = assertThrows(NoSuchUserIdException.class,
+                () -> cacheProxy.getUser(request.userId), "testMaxSizeTest should throw noSuchIdException"
```
```suggestion
                () -> cacheProxy.getUser(request.userId), "testTTL should throw noSuchIdException"
```

kg981167224 commented on user-service/src/main/java/uk/gov/gchq/palisade/service/user/UserApplication.java:
```diff
@@ -88,30 +86,27 @@ public UserApplication(
      */
     public static void main(final String[] args) {
-        LOGGER.debug("UserApplication started with: {} {} {}", UserApplication.class, "main", args);
+        LOGGER.debug("UserApplication started with: {}", (Object) args);
```
Should this be Object[] since you want to call debug(String var1, Object... var2)

Although this change gives a warning, and I think ultimately it will work anyway.

developer01189998819991197253 replied:
It should work fine this way, and matches other services for consistency also

kg981167224 commented on user-service/src/main/java/uk/gov/gchq/palisade/service/user/model/AuditMessage.java:
```diff
@@ -49,28 +49,23 @@
 
     @JsonProperty("resourceId")
-    protected final String resourceId;  //Resource that that is being asked to access.
+    protected final String resourceId;  // Resource that that is being asked to access.
```
Remove blank lines here to match rest of removed lines?

developer01189998819991197253 replied:
removed

kg981167224 commented on user-service/src/main/java/uk/gov/gchq/palisade/service/user/service/KafkaHealthIndicator.java:
```diff
@@ -0,0 +1,101 @@
+
+            if (consumerGroupDescription.state() == ConsumerGroupState.STABLE) {
+                return consumerGroupDescription.members().stream()
+                        .noneMatch(member -> (member.assignment() == null || member.assignment().topicPartitions().isEmpty()));
```
Would a LOGGER.error be good here if we fail and return false?

dev930018 replied:
In the case you return `false`, the service will soon be shut-down by k8s.
Any loggers here would get called periodically (every 20s or so?).
I think a `LOGGER.error("Failed to find kafka topic-partition assignments")` would be reasonable in this case.
Then a `kubectl logs user-service-0 -p` would reveal that there was a kafka problem quite quickly.

developer01189998819991197253 replied:
updated to add a log message

kg981167224 commented on user-service/src/main/java/uk/gov/gchq/palisade/service/user/stream/SerDesConfig.java:
```diff
@@ -108,7 +108,7 @@ private SerDesConfig() {
      * Kafka value serialiser for downstream messages going out as output
      *
-     * @return an appropriate value serialiser for the topic's message content (AttributeMaskingResponse)
+     * @return an appropriate value serialiser for the topic's message content
```
```suggestion
     * @return an appropriate value serializer for the topic's message content
```

developer01189998819991197253 replied:
fixed

kg981167224 commented on user-service/src/main/java/uk/gov/gchq/palisade/service/user/stream/SerDesConfig.java:
```diff
@@ -120,6 +120,15 @@ private SerDesConfig() {
+    /**
+     * Kafka value serialiser for downstream messages going out as output
+     *
+     * @return an appropriate value serialiser for the topic's message content
```
```suggestion
     * @return an appropriate value serializer for the topic's message content
```

developer01189998819991197253 replied:
fixed

kg981167224 commented on user-service/src/main/java/uk/gov/gchq/palisade/service/user/stream/SerDesConfig.java:
```diff
@@ -120,6 +120,15 @@ private SerDesConfig() {
+    }
+
     /**
      * Kafka key serialiser for downstream messages going out as errors
```
```suggestion
     * Kafka key serializer for downstream messages going out as errors
```

developer01189998819991197253 replied:
fixed

kg981167224 reviewed (dismissed):
Nothing urgent so approved with comments.

developer01189998819991197253 committed:
PAL-869, updated README.md to include example objects, and better describe the service.
Updated ApplicationConfiguration to add javadoc and remove loggingErrorHandling
Removed AuditableException.java, updated NoSuchUserIdException and PropertyLoadingException
Added AuditableUserResponse used that in the AkkaRunnableGraph
Removed ErrorHandlingService.java
Added KafkaHealthIndicator.java
Now returning a CompletableFuture in the KafkaProducerService
Updated UserServiceAsyncProxy.java to return a CompletableFuture of AuditableUserResponse that'll contain any error messages if thrown in the service
Added passThroughSerializer to SerDesConfig
Updated UserApplication, and removed discovery annotation
Removed application-eureka.yaml
Updated Caffeine and Redis userCaching tests
Added UserServiceAsyncProxyErrorTest
Updated KafkaContractTest and added ErrorTopicContractTest
Updated HealthActuatorContractTest
Finally updated service image tag

Signed-off-by: developer01189998819991197253 <52958809+developer01189998819991197253@users.noreply.github.com>

dbz120896 commented on user-service/README.md:
```diff
@@ -1,26 +1,104 @@
+    "auths": [
+      "auths"
+    ],
+    "class": "uk.gov.gchq.palisade.User"
```
Should not be returning class. This is exposing the service internal model, which is subject to change.

developer01189998819991197253 replied:
the rest post is an internal api that is only meant for debugging. Inside a cluster these endpoints wont be visible

dbz120896 replied:
Ok. We can let this go then.

dev930018 replied:
If anything, you need this class as it may be a `uk.gov.gchq.palisade.example.ExampleUser` or some other richer object with extra fields

dbz120896 commented on user-service/README.md:
```diff
@@ -1,26 +1,104 @@
 
-Currently the `UserService` supports adding and retrieving of user details via the two following API calls.
+The application exposes two REST endpoints used for debugging or mocking kafka entrypoints:
+* `POST api/user`
```
We are not posting a user, but a request. The endpoint should be `api/users/requests` to keep in line with RESTful semantics.

developer01189998819991197253 replied:
This would require a change to the other services also, as they follow the same convention of /api/*service e.g /api/mask
if its worth doing, we can create a ticket on the backlog

dbz120896 replied:
I think it is worth doing, but is a low priority. An issue in the backlog would be best I think.

dev930018 replied:
I like the usage we currently have as it is meant for pushing items onto kafka queues. In this way, the REST endpoint reflects the name of the topic.

dbz120896 commented on user-service/src/component-tests/java/uk/gov/gchq/palisade/component/user/service/UserServiceAsyncProxyErrorTest.java:
```diff
@@ -0,0 +1,195 @@
+
+
+    private static final Function<Integer, String> REQUEST_FACTORY_JSON = i ->
+            String.format("{\"userId\":\"test-user-id\",\"resourceId\":\"/test/resourceId\",\"context\":{\"class\":\"uk.gov.gchq.palisade.Context\",\"contents\":{\"purpose\":\"purpose\"}}}", i, i);
```
You could do this:
```
String.format("{ 'userId' : 'test-user-id' }").replace("'", "\"");
```
Gets rid of all the unsightly escaping making it a bit easier to read, but is down to preference.

developer01189998819991197253 replied:
We could, but as we show an example of the pretty json in the README, there is no real gain in showing a pretty version here too

dbz120896 reviewed (changes requested):
Only real changes are the class in the response and the URL for the POST endpoint.

developer01189998819991197253 committed:
Merge branch 'develop' into PAL-869-UserService-Auditing

developer01189998819991197253 committed:
PAL-869, Extracted KafkaInitializer into its own class and merged the error and kafka contract tests together to prevent starting kafka twice
Updated the exception message in NoSuchIdException.java

Signed-off-by: developer01189998819991197253 <52958809+developer01189998819991197253@users.noreply.github.com>

dev930018 commented on user-service/README.md:
```diff
@@ -1,26 +1,104 @@
-<span style="color:red">**Note:** As noted in the [documentation root](../README.md) Palisade is in an early stage of development, therefore the precise APIs contained in these service modules are likely to adapt as the project matures.</span>
+The User Service is responsible for providing the other Palisade components with knowledge of the users of the system. 
+Note that we are avoiding having a separate notion of "Palisade" users that are distinct from the normal user accounts of the system. There is no such thing as a "Palisade" user.  
+One of the purposes of this service is to allow Palisade to adopt whatever notion of *user* the host environment has. For example, this may from a central directory service such as LDAP, 
```
```suggestion
One of the purposes of this service is to allow Palisade to adopt whatever notion of *user* the host environment has. For example, this may come from a central directory service such as LDAP, 
```

pd104923 replied:
Done

dev930018 commented on user-service/README.md:
```diff
@@ -1,26 +1,104 @@
+(fields marked with * are acquired from headers metadata)
 
-The user service separates this concern from the rest of the system. Other components use this service's API to request user details. Some deployments may also allow Palisade to add users to the system, hence the presence of the `addUser()` method in the `UserService` interface.
+The service accepts a `UserRequest` and a token, then checks that the User is in the cache; the technology of which is chosen in the User Services application.yaml and Spring profiles.
```
```suggestion
The service accepts a `UserRequest` and a token, then checks if the User is in the cache; the technology of which is chosen in the User Services application.yaml and Spring profiles.
```

pd104923 replied:
Done

dev930018 commented on user-service/README.md:
```diff
@@ -1,26 +1,104 @@
 
-The user service separates this concern from the rest of the system. Other components use this service's API to request user details. Some deployments may also allow Palisade to add users to the system, hence the presence of the `addUser()` method in the `UserService` interface.
+The service accepts a `UserRequest` and a token, then checks that the User is in the cache; the technology of which is chosen in the User Services application.yaml and Spring profiles.
+If the User exists in the cache, then it, alongside any attributes are retrieved, and used to create a `UserReponse` object which is then packaged into a `AuditableUserResponse` object with no errors, 
```
```suggestion
If the User exists in the cache, then it, alongside any attributes are retrieved, and used to create a `UserReponse` object which is then packaged into a `AuditableUserResponse` object with no errors, 
```
```suggestion
If the User exists in the cache, then it, alongside any additional attributes, are retrieved, and used to create a `UserReponse` object which is then packaged into a `AuditableUserResponse` object with no errors.
```

pd104923 replied:
Done

dev930018 commented on user-service/README.md:
```diff
@@ -1,26 +1,104 @@
-The user service separates this concern from the rest of the system. Other components use this service's API to request user details. Some deployments may also allow Palisade to add users to the system, hence the presence of the `addUser()` method in the `UserService` interface.
+The service accepts a `UserRequest` and a token, then checks that the User is in the cache; the technology of which is chosen in the User Services application.yaml and Spring profiles.
+If the User exists in the cache, then it, alongside any attributes are retrieved, and used to create a `UserReponse` object which is then packaged into a `AuditableUserResponse` object with no errors, 
+and then this is sent via kafka, to the Resource Service for further processing. 
```
```suggestion
Then, this is sent via Kafka to the [Resource Service](../resource-service) for further processing. 
```

pd104923 replied:
Done

dev930018 commented on user-service/README.md:
```diff
@@ -1,26 +1,104 @@
+The service accepts a `UserRequest` and a token, then checks that the User is in the cache; the technology of which is chosen in the User Services application.yaml and Spring profiles.
+If the User exists in the cache, then it, alongside any attributes are retrieved, and used to create a `UserReponse` object which is then packaged into a `AuditableUserResponse` object with no errors, 
+and then this is sent via kafka, to the Resource Service for further processing. 
+If the User does not exist in the cache, then a `NoSuchUserIdException` is thrown within the Service, which is caught, and packaged in the `AuditableUserReponse`, alongside the original `UserRequest` and an `AuditErrorMessage`, 
```
If it isn't found in the cache, it calls the real service, which should throw a `NoSuchUserIdException`. Only our 'trivial' implementation, the NullUserService, always throws this exception.

pd104923 replied:
Done

dev930018 commented on user-service/README.md:
```diff
@@ -1,26 +1,104 @@
+If the User exists in the cache, then it, alongside any attributes are retrieved, and used to create a `UserReponse` object which is then packaged into a `AuditableUserResponse` object with no errors, 
+and then this is sent via kafka, to the Resource Service for further processing. 
+If the User does not exist in the cache, then a `NoSuchUserIdException` is thrown within the Service, which is caught, and packaged in the `AuditableUserReponse`, alongside the original `UserRequest` and an `AuditErrorMessage`, 
+this is then sent to the Audit Service and logged appropriatly.
```
```suggestion
this is then sent to the [Audit Service](../audit-service) and audited appropriately.
```

pd104923 replied:
Done

dev930018 commented on user-service/README.md:
```diff
@@ -1,26 +1,104 @@
+## REST Interface
 
-Currently the `UserService` supports adding and retrieving of user details via the two following API calls.
+The application exposes two REST endpoints used for debugging or mocking kafka entrypoints:
```
"exposes two" but have only described one (`api/user/multi` is missing)

pd104923 replied:
Done

dev930018 commented on user-service/src/component-tests/java/uk/gov/gchq/palisade/component/user/service/RedisUserCachingTest.java:
```diff
@@ -0,0 +1,180 @@
+        UserRequest request = UserRequest.Builder.create().withUserId(user.getUserId().getId()).withResourceId("test/resource").withContext(new Context().purpose("purpose"));
+        cacheProxy.addUser(user);
+
+        TimeUnit.MILLISECONDS.sleep(1000);
```
```suggestion
        TimeUnit.SECONDS.sleep(1);
```

pd104923 replied:
Done

dev930018 commented on user-service/src/contract-tests/java/uk/gov/gchq/palisade/contract/user/kafka/KafkaInitializer.java:
```diff
@@ -0,0 +1,155 @@
+
+        // test kafka config
+        String kafkaConfig = "akka.discovery.config.services.kafka.from-config=false";
+        String kafkaPort = "akka.discovery.config.services.kafka.endpoints[0].port" + KAFKA_CONTAINER.getFirstMappedPort();
```
```suggestion
        String kafkaPort = "akka.discovery.config.services.kafka.endpoints[0].port" + KAFKA_CONTAINER.getFirstMappedPort();
```
```suggestion
        String kafkaPort = "akka.discovery.config.services.kafka.endpoints[0].port=" + KAFKA_CONTAINER.getFirstMappedPort();
```

pd104923 replied:
Done

dev930018 commented on user-service/src/component-tests/java/uk/gov/gchq/palisade/component/user/service/UserServiceAsyncProxyErrorTest.java:
```diff
@@ -0,0 +1,196 @@
+    private static final Logger LOGGER = LoggerFactory.getLogger(UserServiceAsyncProxyErrorTest.class);
+
+
+    private static final Function<Integer, String> REQUEST_FACTORY_JSON = i ->
```
I get why we'd have this in contract tests, but why here?

pd104923 replied:
This isn't actually used properly as we return the same json String each time. Adjusted to just be a `String` value.

dev930018 commented on user-service/src/component-tests/java/uk/gov/gchq/palisade/component/user/service/UserServiceAsyncProxyErrorTest.java:
```diff
@@ -0,0 +1,196 @@
+                .isEqualTo(NoSuchUserIdException.class.getName() + ": No userId matching test-user-id found in cache");
+    }
+
+    public static class RedisInitializer implements ApplicationContextInitializer<ConfigurableApplicationContext> {
```
This shouldn't need redis. Use caffeine if you need a cache implementation for this test

pd104923 replied:
Reworked this class to use Caffeine

dev930018 commented on user-service/src/contract-tests/java/uk/gov/gchq/palisade/contract/user/kafka/KafkaContractTest.java:
```diff
@@ -230,7 +169,7 @@ void testRequestSet() {
                                         .isEqualTo(ContractTestData.REQUEST_TOKEN.getBytes())),
                 () -> assertThat(results.get(0).value().get("user").get("userId").get("id").asText())
-                        .isEqualTo(ContractTestData.USER_ID.getId()),
+                        .isEqualTo("test-user-id"),
```
Suggest reverting to use ContractTestData

pd104923 replied:
Reverted

dev930018 commented on user-service/src/contract-tests/java/uk/gov/gchq/palisade/contract/user/rest/HealthActuatorContractTest.java:
```diff
@@ -33,8 +34,12 @@
+@SpringBootTest(
+        classes = UserApplication.class,
+        webEnvironment = WebEnvironment.RANDOM_PORT,
+        properties = {"management.health.kafka.enabled=false", "spring.autoconfigure.exclude=org.springframework.boot.autoconfigure.data.redis.RedisAutoConfiguration"}
```
This Redis exclusion shouldn't be required if using the caffeine profile

pd104923 replied:
Removed the exclusions

dev930018 commented on user-service/src/main/java/uk/gov/gchq/palisade/service/user/model/AuditMessage.java:
```diff
@@ -49,28 +49,21 @@
+    protected final String serverHostname;  // The hostname of the server hosting the service
     @JsonProperty("attributes")
-    protected final Map<String, Object> attributes;  //Map<String, Object> holding optional extra information
+    protected final Map<String, Object> attributes;  // Map<String, Object> holding optional extra information
```
At some point, it's probably worth changing this, but you decide if wanted now
```suggestion
    protected final Map<String, String> attributes;  // Map<String, String> holding optional extra information
```

pd104923 replied:
I think this should be left for now so that it's done on all the services at the same time.

dev930018 commented on user-service/src/main/java/uk/gov/gchq/palisade/service/user/service/KafkaHealthIndicator.java:
```diff
@@ -0,0 +1,106 @@
+ * Kafka health indicator. Check that the consumer group can be accessed and is registered with the cluster,
+ * if not mark the service as unhealthy.
+ */
+@Component("kafka")
+@ConditionalOnEnabledHealthIndicator("kafka")
+public class KafkaHealthIndicator implements HealthIndicator {
```
Prefer explicit `@Bean` in `@Configuration` over `@Component`

pd104923 replied:
A bean has been added to the ApplicationConfiguration.

dev930018 commented on user-service/src/main/java/uk/gov/gchq/palisade/service/user/service/KafkaHealthIndicator.java:
```diff
@@ -0,0 +1,106 @@
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(KafkaHealthIndicator.class);
+    private final AdminClient adminClient;
+    @Value("${akka.kafka.consumer.kafka-clients.group.id}")
+    private String groupId;
```
Prefer taking the config object (see somewhere in the `stream` package) as an argument to the constructor over using the `@Value` annotation

pd104923 replied:
The topic configuration has been added to the constructor for this class. The "groupId" field is set within the `performCheck` method.

dev930018 commented on user-service/src/main/java/uk/gov/gchq/palisade/service/user/service/KafkaHealthIndicator.java:
```diff
@@ -0,0 +1,106 @@
+        try {
+            Map<String, ConsumerGroupDescription> groupDescriptionMap = this.adminClient.describeConsumerGroups(Collections.singletonList(this.groupId))
+                    .all()
+                    .get(1, TimeUnit.SECONDS);
```
```suggestion
                    .join();
```
avoids the need for the catch

pd104923 replied:
The `.join()` method is not available to use here as it's a `KafkaFuture`

dev930018 commented on user-service/src/main/java/uk/gov/gchq/palisade/service/user/service/KafkaHealthIndicator.java:
```diff
@@ -0,0 +1,106 @@
+                }
+                return assignedGroupPartition;
+            }
+        } catch (InterruptedException | ExecutionException | TimeoutException e) {
```
Suggest us COmpletableFuture::join and avoid the need for this catch

pd104923 replied:
These are necessary as we're dealing with `KafkaFuture`s, not `CompletableFuture`s.

dev930018 commented on user-service/src/main/java/uk/gov/gchq/palisade/service/user/service/KafkaHealthIndicator.java:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    @Override
+    public Health getHealth(final boolean includeDetails) {
+        return Optional.of(performCheck())
+                .filter(healthy -> healthy)
+                .map(up -> Health.up().withDetail("group", this.groupId).build())
+                .orElseGet(() -> Health.down().withDetail("group", this.groupId).build());
+    }
```
There's a lot more detail we *could* be adding here, but i think groupId is sufficient

pd104923 replied:
I think groupId is useful for now but you're right, we could add a lot more information here if required.

dev930018 commented on user-service/src/main/java/uk/gov/gchq/palisade/service/user/service/KafkaHealthIndicator.java:
```diff
@@ -0,0 +1,106 @@
+            Thread.currentThread().interrupt();
+            return false;
+        }
+        return true;
```
By default it's healthy? Hmm...
```suggestion
        return false;
```

pd104923 replied:
The logic for this has been changed slightly so false is returned unless the try is successful.

dev930018 commented on user-service/src/main/java/uk/gov/gchq/palisade/service/user/service/KafkaHealthIndicator.java:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    @Override
+    public Health getHealth(final boolean includeDetails) {
+        return Optional.of(performCheck())
+                .filter(healthy -> healthy)
+                .map(up -> Health.up().withDetail("group", this.groupId).build())
+                .orElseGet(() -> Health.down().withDetail("group", this.groupId).build());
+    }
```
This looks a bit weird, and the Optional is never empty so why use it? Why not:
```
if (performCheck()) {
  return Health.up().with...
} else {
  return Health.down().with...
}

pd104923 replied:
Changed

dev930018 commented on user-service/src/main/java/uk/gov/gchq/palisade/service/user/service/KafkaProducerService.java:
```diff
@@ -92,15 +90,21 @@ public KafkaProducerService(final Sink<ProducerRecord<String, UserRequest>, Comp
-                .toCompletableFuture()
-                .join();
 
-        // Return results
-        return new ResponseEntity<>(HttpStatus.ACCEPTED);
+                // Return a CompletableFuture<Void> result
+                .toCompletableFuture()
+                .thenApply(x -> null);
```
I assume returning a reasonable status code etc. is still done in the controller?

pd104923 replied:
Correct, the controller returns a `202` status code.

dev930018 reviewed (changes requested):
43/63 files viewed, will complete tomorrow

dev930018 commented on user-service/src/main/java/uk/gov/gchq/palisade/service/user/service/UserServiceCachingProxy.java:
```diff
@@ -28,25 +28,40 @@
+     * Default constructor used to create the UserServiceCachingProxy
      *
-     * @param service an implementation of the {@link UserService}
+     * @param service {@link UserService} this service calls
```
I prefer the original documentation

pd104923 replied:
Reverted

dev930018 commented on user-service/src/main/java/uk/gov/gchq/palisade/service/user/service/UserServiceCachingProxy.java:
```diff
@@ -28,25 +28,40 @@
 
+    /**
+     * Using the userId as the key, retrieves the {@link User} from the cache, if the User doesn't exist,
+     * then a message is Logged, and a {@link NullUserService} getUser() is called,
+     * which will throw a {@link uk.gov.gchq.palisade.service.user.exception.NoSuchUserIdException}.
```
While this is true at the moment, it isn't in general: the service is called, which may or may not throw a NoSuchUserIdException

pd104923 replied:
Generalised the JavaDoc

dev930018 commented on user-service/src/main/java/uk/gov/gchq/palisade/service/user/stream/PropertiesConfigurer.java:
```diff
@@ -54,12 +54,12 @@
  */
 public class PropertiesConfigurer extends PropertySourcesPlaceholderConfigurer implements InitializingBean {
-    private static final Pattern INDEXED_PROPERTY_PATTERN = Pattern.compile("^\\s*(?<path>\\w+(?:\\.\\w+)*)\\[(?<index>\\d+)\\]\\.*(.*?)$");
+    private static final Pattern INDEXED_PROPERTY_PATTERN = Pattern.compile("^\\s*(?<path>\\w+(?:\\.\\w+)*)\\[(?<index>\\d+)]\\.*(.*?)$");
```
Are you sure? We're trying to match a literal `[index]` so this `]` should be escaped as it was

pd104923 replied:
Reverted

dev930018 commented on user-service/src/main/resources/application.yaml:
```diff
@@ -50,17 +42,15 @@ management:
-        loggerLevel: basic
-    default-config: default
-
-user-service:
-  implementation: null-user-service
```
Do we no longer need this selector?

pd104923 replied:
I think we still need this so have added it back in.

pd104923 committed:
Pal 869 updated README pr comments

pd104923 committed:
Pal 869 addressed more pr comments

pd104923 committed:
Pal 869 addressed further pr comments

dev930018 commented on user-service/README.md:
```diff
@@ -1,26 +1,111 @@
+The application exposes two REST endpoints used for debugging or mocking kafka entrypoints:
+* `POST api/user`
+  - accepts an `x-request-token` `String` header, any number of extra headers, and a single `UserRequest` in the body
+  - returns a `202 ACCEPTED` after writing the headers and `AuditableUserResponse` to kafka
+* `POST api/user/multi`
+  - accepts an `x-request-token` `String` header, any number of extra headers, and a list of `UserRequest`s within the body
+  - returns a `202 ACCEPTED` after writing the headers and `AuditableUserResponse` to kafka
```
It writes a UserRequest to kafka and returns, no?

pd104923 replied:
You're quite right. Updated so that a UserRequest is sent to Kafka :D

dev930018 commented on user-service/src/component-tests/java/uk/gov/gchq/palisade/component/user/service/UserServiceAsyncProxyErrorTest.java:
```diff
@@ -0,0 +1,149 @@
+@ActiveProfiles({"caffeine"})
+class UserServiceAsyncProxyErrorTest {
+
+    private static final String REQUEST_JSON = "{\"userId\":\"test-user-id\",\"resourceId\":\"/test/resourceId\",\"context\":{\"class\":\"uk.gov.gchq.palisade.Context\",\"contents\":{\"purpose\":\"purpose\"}}}";
+    private static final String NO_USER_REQUEST_JSON = "{\"userId\":\"Not-a-real-user\",\"resourceId\":\"/test/resourceId\",\"context\":{\"class\":\"uk.gov.gchq.palisade.Context\",\"contents\":{\"purpose\":\"purpose\"}}}";
```
Why have this in a component test? Just create the objects, we don't care about the json

pd104923 replied:
Removed all the JSON stuff. Just creates the objects.

dev930018 commented on user-service/src/component-tests/java/uk/gov/gchq/palisade/component/user/service/UserServiceAsyncProxyErrorTest.java:
```diff
@@ -0,0 +1,149 @@
+                .isNotNull();
+
+        // Then check the error message contains the correct message
+        assertThat(auditableUserResponse.getAuditErrorMessage().getError().getMessage())
+                .as("verify that exception is propagated into an auditable object and returned")
+                .isEqualTo(NoSuchUserIdException.class.getName() + ": No userId matching Not-a-real-user found in cache");
+    }
```
I'd prefer to check the `assertThat(error).hasType(NoSuchUserIdException.class)` rather than inspecting the `getMessage` string

pd104923 replied:
The issue with this is that the "Throwable" error is actually a `CompletionException` with a message that contains the NoSuchUserIdException class.
Hence why we have the classname and the message in the `isEqualTo` check.

dev930018 commented on user-service/src/contract-tests/java/uk/gov/gchq/palisade/contract/user/kafka/KafkaContractTest.java:
```diff
@@ -271,36 +213,56 @@ void testNoSuchUserIdExceptionIsThrown() {
+        assertAll("Asserting on the error topic",
+                // One error is produced
+                () -> assertThat(errorResults)
+                        .hasSize(1),
 
-                () -> assertThat(results.getLast().headers().lastHeader(StreamMarker.HEADER).value())
-                        .isEqualTo(StreamMarker.END.toString().getBytes())
-        );
+                // The error has the relevant headers, including the token
+                () -> assertThat(errorResults)
+                        .allSatisfy(result ->
+                                assertThat(result.headers().lastHeader(Token.HEADER).value())
+                                        .isEqualTo(ContractTestData.REQUEST_TOKEN.getBytes())),
```
Combine these two related assertions:
```
assertThat(errorResults)
  .hasSize(1)
  .allSatisfy(result -> assertThat(headers are correct))
  .allSatisfy(result -> assertThat(value.get("error").get("message")).isEqualTo(...))
```

pd104923 replied:
Updated

dev930018 commented on user-service/src/contract-tests/java/uk/gov/gchq/palisade/contract/user/kafka/KafkaContractTest.java:
```diff
@@ -271,36 +213,56 @@ void testNoSuchUserIdExceptionIsThrown() {
+                // The error has a message that contains the throwable exception, and the message
+                () -> assertThat(errorResults.get(0).value().get("error").get("message").asText())
+                        .isEqualTo(NoSuchUserIdException.class.getName() + ": No userId matching invalid-user-id found in cache")
+        );
```
Ditto, combine this with the above

pd104923 replied:
Done

dev930018 reviewed (changes requested):
Also need to clean up the pom a bit:
* remove embedded-redis in favour of testcontainers 
* remove spring-cloud-config-client and spring-boot-configuration-processor (i think they were used for feign/eureka stuff?)
* remove spring-boot-starter-jpa and h2
*

pd104923 committed:
Pal 869 updarted README, refactored kafka test assertions, updated proxyError tests

pd104923 committed:
Pal 869 adjusted proxy error test assertion

developer01189998819991197253 committed:
Merge branch 'develop' into PAL-869-UserService-Auditing

pd104923 committed:
Merge branch 'develop' into PAL-869-UserService-Auditing

pd104923 committed:
Pal 949 initial smell update, changed Websocket values to WebSocket

pd104923 committed:
Pal 949 added JavaDoc to AuditableWebSocketMessage

pd104923 committed:
Pal 949 resolved sq smells

pd104923 committed:
Pal 949 fixed checkstyle issue

developer01189998819991197253 committed:
Merge branch 'develop' into PAL-869-UserService-Auditing

pd104923 committed:
Merge branch 'develop' into Pal-949-sq-develop-smells

developer01189998819991197253 committed:
PAL-949 updated WebSocketMessage variables within Filtered Resource Service to use camel case appropriatley

developer01189998819991197253 committed:
Merge branch 'Pal-949-sq-develop-smells' into PAL-869-UserService-Auditing

ac74475 committed:
Merge branch 'develop' into PAL-869-UserService-Auditing

ac74475 commented:
possible pipeline race condition being looked into, hold merge until it has been checked out

dev-958 committed:
Merge branch 'develop' into PAL-869-UserService-Auditing
